spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/core/PubSubTemplate.java

```diff
@@ -40,39 +43,65 @@
 import org.springframework.beans.factory.InitializingBean;
 import org.springframework.cloud.gcp.pubsub.support.PublisherFactory;
 import org.springframework.cloud.gcp.pubsub.support.SubscriberFactory;
+import org.springframework.integration.support.json.JacksonJsonUtils;
 import org.springframework.util.Assert;
 import org.springframework.util.concurrent.ListenableFuture;
 import org.springframework.util.concurrent.SettableListenableFuture;
 
 /**
+ * // @formatter:off
  * Default implementation of {@link PubSubOperations}.
  *
- * <p>The main Google Cloud Pub/Sub integration component for publishing to topics and consuming messages from
- * subscriptions asynchronously or by pulling.
- *
+ * <p>The main Google Cloud Pub/Sub integration component for publishing to topics and
+ * consuming messages from subscriptions asynchronously or by pulling.
+ * // @formatter:on
  * @author Vinicius Carvalho
  * @author João André Martins
  * @author Mike Eltsufin
+ * @author Chengyuan Zhao
  */
 public class PubSubTemplate implements PubSubOperations, InitializingBean {
 
 	private static final Log LOGGER = LogFactory.getLog(PubSubTemplate.class);
 
+	private final ObjectMapper objectMapper;
+
 	private final PublisherFactory publisherFactory;
 
 	private final SubscriberFactory subscriberFactory;
 
 	/**
-	 * Default {@link PubSubTemplate} constructor.
+	 * {@link PubSubTemplate} constructor.
 	 *
 	 * @param publisherFactory the {@link com.google.cloud.pubsub.v1.Publisher} factory to
-	 *                         publish to topics
-	 * @param subscriberFactory the {@link com.google.cloud.pubsub.v1.Subscriber} factory to
-	 *                          subscribe to subscriptions
+	 * publish to topics
+	 * @param subscriberFactory the {@link com.google.cloud.pubsub.v1.Subscriber} factory
+	 * to subscribe to subscriptions
+	 * @param trustedPackages the trusted Java packages for deserialization. If null or
+	 * empty, then default type information is not included in serialization.
 	 */
-	public PubSubTemplate(PublisherFactory publisherFactory, SubscriberFactory subscriberFactory) {
+	public PubSubTemplate(PublisherFactory publisherFactory,
+			SubscriberFactory subscriberFactory, String[] trustedPackages) {
 		this.publisherFactory = publisherFactory;
 		this.subscriberFactory = subscriberFactory;
+		this.objectMapper = trustedPackages == null || trustedPackages.length == 0
+				? new ObjectMapper()
```

Comment:
I think the creation of the `ObjectMapper` should be completely contained in the auto-config class. You shouldn't need to call `new` here or pass in the `trustedPackages`.

Reply:
Could you elaborate on that? I think the logic of "if no trusted packages are specified , then don't include type information in JSON" is very specific to this specific class.

Reply:
??? Why don't always include type information? 
It is very convenient way do not worry about type on the other side, especially when you deal with container classes, like `List` or `Message`.

To trust or not trust is already an end-user choice and his/her security team.
However we can have an opinion on the matter and always be secured providing some default set of packages. For example in Spring Integration it is:
```
private static final List<String> TRUSTED_PACKAGES =
			Arrays.asList(
					"java.util",
					"java.lang",
					"org.springframework.messaging.support",
					"org.springframework.integration.support",
					"org.springframework.integration.message",
					"org.springframework.integration.store"
			);
```

To *trust all* we just require an asterisk (`*`)":
```
if (trustedPackages != null) {
		for (String whiteListClass : trustedPackages) {
			if ("*".equals(whiteListClass)) {
				this.trustedPackages.clear();
				break;
			}
			else {
				this.trustedPackages.add(whiteListClass);
			}
		}
}
```

So, this way you will never use `ObjectMapper` without `trustedPackages` setting. However I would still defer the choice to end user and would let him/her to inject any possible `ObjectMapper`, but that's already a `MessageConverter` (or `PayloadConverter`) responsibility.

Reply:
If the user doesn't provide any trusted packages, then is that supposed to mean "trust all" ?

Reply:
In Spring Kafka we use `"java.util", "java.lang"` as default trusted packages.